### DISCUSSION TITLE
The database schema

### DISCUSSION THREAD

**Discussion by JanSkoda** [adminerevo/adminerevo#90] [Ideas]
Is there a chance to improve rendering a database schema little bit?
I am using it for reporting the status and visualising the relations for my colleagues, which are developing the devices connecting to my server. When i print the schema, the connection lines are sometimes weirdly merging and crossing and it makes it a bit unreadable.

**Comment by LionelLaffineur** [comment]
Dear @JanSkoda, I don't think that kind of schema (which can be very wide and complex) is well the business of AdminerEvo.
To follow your suggestion, I'm trying to get in touch with someone at Draw.io to see if it would be possible to generate the schemas from within AdminerEvo in a format to view/edit in Draw.io.
I think it could be an improvement because it would leave the layout to an application for which it is the core business.

**Comment by peterdd** [comment]
I once ~~hacked~~ made improvements to the schema, see my fork www.github.com/peterdd/adminer

It has only some primitive layout algorithms, but better than nothing.

Screenshots taken after importing a gitlab structure.sql as example and little stress test.

The drivers/pgsql.inc.php of my 5 year old version is outdated, but the current vrana/adminer drivers/pgsql.inc.php is also broken. After patching the foreign key part it is able to render >700 tables with foreign key constraints, although my browser, CPU and fan do not liked it. ( per line for example)

Screenshots taken from after importing a gitlab structure.sql.

The upper diagram contains only tables starting with 'user' (`substr($table,0,4)=='user'`), so some connections point to nothing(0,0). Manual positioning.(a bit clunky drag&drop & using existing cookie storage)

Below shows all tables of default gitlab with all foreign key relations using a bit different method of line coloring.

![peterdd_adminer_db_schema_diagram](https://github.com/adminerevo/adminerevo/assets/1839154/ab3e9202-7b91-488b-adb4-d3cfe37c1f17)

**Reply by LionelLaffineur** [reply]
Dear @peterdd, thanks for sharing with us.  May I ask you how I could do to test it myself ? Regards, L.

**Reply by peterdd** [reply]
I used git version of http://www.github.com/peterdd/adminer

and current www.github/vrana/adminer/driver.pgsql.inc.php

and reverting 

https://github.com/vrana/adminer/commit/60ad16117868103cb053881ecf14303236b6bcbc

That idf_escape() function made garbage output with the function foreign_keys() for me.

I just not did any compiling or minifying (I even not know what that jush stuff is for and not feel the 'everything minified into 1 file' philosophy), but I only used it locally for testing. Just download current current master as zip, unzip in some webdir, patch pgsql.inc.php and open in browser and connect to a prepared pg database with some tables.

For the example screenshots I imported the structure.sql from https://gitlab.com/gitlab-org/gitlab/-/blob/master/db/structure.sql into a demo postgresql db and connected mit adminer to it.

**Reply by peterdd** [reply]
The minimap approach IMHO only works with Firefox as the other browser vendors not implemented the feature I used.
